Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2022, Leon Albrecht <leon.a@serenityos.com>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Bitmap.h>
+#include <AK/TypeCasts.h>
+#include <LibJS/Bytecode/Op.h>
+#include <LibJS/Bytecode/PassManager.h>
+
+namespace JS::Bytecode::Passes {
+
+static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t number_of_registers)
+{
+    auto array_ranges = Bitmap::must_create(number_of_registers, false);
+
+    for (auto it = InstructionStreamIterator(block.instruction_stream()); !it.at_end(); ++it) {
+        if ((*it).type() == Instruction::Type::NewArray) {
+            Op::NewArray const& array_instruction = static_cast<Op::NewArray const&>(*it);
+            if (size_t element_count = array_instruction.element_count())
+                array_ranges.set_range<true, false>(array_instruction.start().index(), element_count);
+        }
+    }
+
+    auto new_block = BasicBlock::create(block.name(), block.size());
+    HashMap<size_t, Register> identifier_table {};
+    HashMap<u32, Register> register_rerouteing_table {};
+
+    for (auto it = InstructionStreamIterator(block.instruction_stream()); !it.at_end();) {
+        using enum Instruction::Type;
+        switch ((*it).type()) {
+        case GetVariable: {
+            auto const& get_variable = static_cast<Op::GetVariable const&>(*it);
+            ++it;
+            auto const& next_instruction = *it;
+
+            if (auto reg = identifier_table.find(get_variable.identifier().value()); reg != identifier_table.end()) {
+                // When we have already seen a variable, we can replace its GetVariable with a simple Load
+                // knowing that it was already stored in a register
+                new (new_block->next_slot()) Op::Load(reg->value);
+                new_block->grow(sizeof(Op::Load));
+
+                if (next_instruction.type() == Instruction::Type::Store) {
+                    // If the next instruction is a store, we can eliminate that store
+                    // with a rerouteing all instructions using the destination of that store
+                    // This only works, if the destination is not part of an array construction,
+                    // because these need a contigious range of registers,
+                    // in that case we re-emit the store
```

Comment:
This could use a second look re typos and flow, it's quite confusing to read.

Reply:
Rephrased that, I hope its easier to understand noe